### PR TITLE
[react-native] Update Typings for 0.65

### PR DESCRIPTION
**Pull request by NickGerleman**
This is a best effort update of typings by going through https://github.com/facebook/react-native/compare/0.64-stable...0.65-stable then looking at any changes to Flow typings in the Libraries directory that may be public.

We version the existing typings at the same time to a v0.64 directory.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [ ] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [ ] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [ ] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<https://github.com/facebook/react-native/compare/0.64-stable...0.65-stable>>
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Review by elicwhite — APPROVED** [review]
dogscience

**Comment by efstathiosntonas** [comment]
@NickGerleman  Any chance to include https://github.com/facebook/react-native/issues/32054 fix? Thanks

**Review by kuasha420 — APPROVED** [review]

**Comment by JonnyBurger** [comment]
@NickGerleman The listener types for `Linking` should also be updated, to something like this:

```diff
export interface LinkingStatic extends NativeEventEmitter {
    /**
     * Add a handler to Linking changes by listening to the `url` event type
     * and providing the handler
     */
+   addEventListener(type: 'url', handler: (event: { url: string }) => void): EmitterSubscription;
-   addEventListener(type: 'url', handler: (event: { url: string }) => void): void;

    /**
     * Remove a handler by passing the `url` event type and the handler
+    * @deprecated Call remove() on the return value of addEventListener() instead.
     */
    removeEventListener(type: 'url', handler: (event: { url: string }) => void): void;
```

**Comment by gantonious** [comment]
Looks like some `Dimensions` needs some updates as well. `addEventListener` now returns an `EventSubscription` and `removeEventListener` is deprecated.

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/28ff99ff3410db3be8d27f231a5b9a14b6ddf89e/types/react-native/index.d.ts#L6150-L6164

`react-native` diff:

```diff
@@ -100,17 +102,20 @@ class Dimensions {
    *   are the same as the return values of `Dimensions.get('window')` and
    *   `Dimensions.get('screen')`, respectively.
    */
-  static addEventListener(type: 'change', handler: Function) {
+  static addEventListener(
+    type: 'change',
+    handler: Function,
+  ): EventSubscription {
     invariant(
       type === 'change',
       'Trying to subscribe to unknown event: "%s"',
       type,
     );
-    eventEmitter.addListener(type, handler);
+    return eventEmitter.addListener(type, handler);
   }
 
   /**
-   * Remove an event handler.
+   * @deprecated Use `remove` on the EventSubscription from `addEventListener`.
    */
   static removeEventListener(type: 'change', handler: Function) {
     invariant(
@@ -118,6 +123,7 @@ class Dimensions {
       'Trying to remove listener for unknown event: "%s"',
       type,
     );
+    // NOTE: This will report a deprecation notice via `console.error`.
     eventEmitter.removeListener(type, handler);
   }
 }
```

Diff generated using:
```
git diff v0.64.2..v0.65.0 -- Libraries/Utilities/Dimensions.js
```

**Comment by NickGerleman** [comment]
@JonnyBurger @gantonious updated

File: types/react-native/index.d.ts
Comment (robinvw1) [inline types/react-native/index.d.ts:7404]: Instead of removing, can we mark it as `@deprecated` just like you did [here](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/55354/files#diff-e94ad9b79ba4f07235bfefa12c2789cd2d6cb690d9120309555f33ae51295244R7709)?

**Review by orta — APPROVED** [review]
Looks reasonable to me

File: types/react-native/index.d.ts
Comment (NickGerleman) [reply types/react-native/index.d.ts:7404]: I totally missed this comment last week. I was under the impression the method was removed, but you're right. It looks like it was restored in a later PR. https://github.com/facebook/react-native/commit/035718ba97bb44c68f2a4ccdd95e537e3d28690c#diff-3402d8d0082acba7a28c141bf8ca9aa3ae63c4dd9dfcf210c3d4b5bea739bdc1

I'd normally make this change right now, but am on vacation this week, and am not sure I will be able to quickly follow up once sign off happens again. It's been a couple of weeks since 0.65 has been out, so I think it would be better to cut the typings as-is to unblock adoption, with 0.66 coming very soon.

I'm going to merge this change as-is, but am keeping this in mind as a followup for when I am back. Folks have been great on this thread in finding fixups, and I am hoping merging what is here now as 0.65 should unblock folks PR'ing them directly.

**Comment by NickGerleman** [comment]
Ready to merge

**Pull request merged**
